akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -366,23 +364,28 @@ import scala.util.control.NonFatal
         }
       }
 
-      def enqueueAndNotify(requested: Received[T]): Unit = {
-        buffer.enqueue(requested)
-        currentRequest match {
-          case Some(p) =>
-            sendDownstream(p)
-            currentRequest = None
-          case None => //do nothing
-        }
-      }
-
       def onPush(): Unit = {
-        enqueueAndNotify(Success(Some(grab(in))))
+        buffer.enqueue(Success(Some(grab(in))))
+        if (currentRequests.nonEmpty) currentRequests.dequeue().complete(buffer.dequeue())
         if (buffer.used < maxBuffer) pull(in)
       }
 
-      override def onUpstreamFinish(): Unit = enqueueAndNotify(Success(None))
-      override def onUpstreamFailure(ex: Throwable): Unit = enqueueAndNotify(Failure(ex))
+      override def onUpstreamFinish(): Unit = {
+        buffer.enqueue(Success(None))
+        while (currentRequests.nonEmpty && buffer.nonEmpty) currentRequests.dequeue().complete(buffer.dequeue())
+        while (currentRequests.nonEmpty) currentRequests.dequeue().complete(Success(None))
+        if (buffer.isEmpty) completeStage()
```

Comment:
when/how will the stage be completed if this `if` is false?

Reply:
When subscribers register they will immediately consume what's in the buffer, if it reaches the Success(None) of Failure(ex) it will complete/fail the stage.

I wonder though for failure if it shouldn't just fail immediately.

Reply:
These are the handles for the upstream, elements that are already passed to the queue should not depend from upstream and we allow the queue to keep going until there are elements in the buffer.